app/database/result_store_propagate.go

```diff
@@ -0,0 +1,292 @@
+package database
+
+import (
+	"database/sql"
+	"time"
+)
+
+const propagateLockName = "listener_propagate"
+const propagateLockTimeout = 10 * time.Second
+
+// Propagate recomputes fields of results
+// For results marked with result_propagation_state = 'to_be_propagated'/'to_be_recomputed':
+// 1. We mark all their ancestors in results as 'to_be_recomputed'
+//  (we consider a row in results as an ancestor if it has the same value in group_id and
+//  its item_id is an ancestor of the original row's item_id).
+// 2. We process all objects that are marked as 'to_be_recomputed' and that have no children marked as 'to_be_recomputed'.
+//  Then, if an object has children, we update
+//    latest_activity_at, tasks_tried, tasks_with_help, validated_at.
+//  This step is repeated until no records are updated.
+// 3. We insert new permissions_granted for each unlocked item according to corresponding item_unlocking_rules.
+func (s *ResultStore) Propagate() (err error) {
+	s.mustBeInTransaction()
+	defer recoverPanics(&err)
+
+	var groupsUnlocked int64
+
+	// Use a lock so that we don't execute the listener multiple times in parallel
+	mustNotBeError(s.WithNamedLock(propagateLockName, propagateLockTimeout, func(ds *DataStore) error {
+		// We mark as 'to_be_recomputed' results of all ancestors of items marked as 'to_be_recomputed'/'to_be_propagated'
+		// with appropriate attempt_id.
+		// Also, we insert missing results for chapters having descendants with results marked as 'to_be_recomputed'/'to_be_propagated'.
+		// We only create results for chapters which are (or have ancestors which are) visible to the group that attempted
+		// to solve the descendant items. Chapters requiring explicit entry or placed outside of the scope
+		// of the attempts's root item are skipped).
+		// (This query can take more than 30 seconds to run when executed for the first time after the db migration)
+		mustNotBeError(ds.Exec(`
```

Comment:
Wow, even if I get the idea, it is impossible to say if this request is correct and if it has a chance to scale. :-/

Reply:
I'm not sure I understand what you mean. It is possible to say if the query is correct since we have tests: if the tests pass, this means the query is correct. The new query is not much more complex than the old one and  it works great (and even faster than the old one) on the example DB.

Reply:
Yes but I try to base myself on the query. Seems ok but I cannot convince myself at 100%.
Ok for perf. 👍